cn-Introducing-JSX (#15)
reactjs/zh-hans.reactjs.org: content/docs/introducing-jsx.md

---
id: introducing-jsx
title: JSX 简介
permalink: docs/introducing-jsx.html
prev: hello-world.html
next: rendering-elements.html
---

考虑如下声明变量：

```js
const element = <h1>Hello, world!</h1>;
```

这个有趣的标签语法既不是字符串也不是 HTML。

> 它被称为 JSX， 是一个 JavaScript 的语法扩展。我们建议在 React 中配合使用 JSX，JSX 可以很好地描述 UI 应该呈现出它应有交互的本质形式。JSX 可能使人想起模版语言，但它具有 JavaScript 的全部功能。

Reviewer: ```suggestion
它被称为 JSX，是一个 JavaScript 的语法扩展。我们建议在 React 中配合使用 JSX，JSX 可以很好地描述 UI 应该呈现出它应有交互的本质形式。JSX 可能使人想起模版语言，但它具有 JavaScript 的全部功能。
```

React 不一定要使用 JSX (/docs/react-without-jsx.html)，但是大多数人发现在JavaScript 代码中使用 UI 时它是一种有用的视觉辅助工具。它还允许 React 显示更多有用的错误和警告消息。

弄清楚这个，让我们开始吧！

### 在 JSX 中嵌入表达式

在下面的例子中，我们声明了一个名为 `name` 的变量，然后在 JSX 中使用它，并将它包装在大括号中：

```js{1,2}
const name = 'Josh Perez';
const element = <h1>Hello, {name}</h1>;

ReactDOM.render(
element,
document.getElementById('root')
);
```

你可以在 JSX 中的大括号内放置任何有效的[JavaScript 表达式] (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Expressions_and_Operators#Expressions)。例如，`2 + 2`，`user.firstName`或`formatName（user）`都是有效的 JavaScript 表达式。

在下面的示例中，我们将调用 JavaScript 函数 `formatName（user）` 的结果， 并将结果嵌入到 `<h1>` 元素中。

```js{12}
function formatName(user) {
return user.firstName + ' ' + user.lastName;
}

const user = {
firstName: 'Harper',
lastName: 'Perez'
};

const element = (
<h1>
Hello, {formatName(user)}!
</h1>
);

ReactDOM.render(
element,
document.getElementById('root')
);
```

[在codepen上运行](codepen://introducing-jsx)

为了便于阅读，我们将 JSX 拆分为多行。 虽然不需要这样，但在执行此操作时，我们还是建议将其包装在括号中，以避免掉入[分号自动插入]的陷阱 (http://stackoverflow.com/q/2846283)。

### JSX 本身也是一种表达式

编译之后，JSX 表达式成为常规 JavaScript 函数调用， 并评估为 JavaScript 对象。

> 这意味着你可以在 if 语句和 for 循环中使用 JSX，将其赋值给变量，当作参数传入，并从函数返回它：

Reviewer: ```suggestion
这意味着你可以在 `if` 语句和 `for` 循环中使用 JSX，将其赋值给变量，当作参数传入，并从函数返回它：
```

```js{3,5}
function getGreeting(user) {

> if (user) {

Reviewer: ```suggestion
  if (user) {
```

> return <h1>Hello, {formatName(user)}!</h1>;

Reviewer: ```suggestion
    return <h1>Hello, {formatName(user)}!</h1>;
```

> }

Reviewer: ```suggestion
  }
```

> return <h1>Hello, Stranger.</h1>;

Reviewer: ```suggestion
  return <h1>Hello, Stranger.</h1>;
```

}
```

> ### JSX 属性

Reviewer: ```suggestion
### JSX 属性详解
```

你可以使用引号来定义以字符串为值的属性：

```js
const element = <div tabIndex="0"></div>;
```

也可以使用大括号来定义以 JavaScript 表达式为值的属性：

```js
const element = <img src={user.avatarUrl}></img>;
```

> 切记，如果把 JavaScript 表达式镶嵌在大括号里，大括号外面不能再套引号。JSX 会将引号当中的内容识别为字符串而不是表达式。你可以只使用引号 （当对象是字符串），或者使用大阔号（当对象是表达式），但这两个不能在同一个属性里出现。

Reviewer: ```suggestion
切记，如果把 JavaScript 表达式镶嵌在大括号里，大括号外面不能再套引号。JSX 会将引号当中的内容识别为字符串而不是表达式。你可以只使用引号（当对象是字符串），或者使用大阔号（当对象是表达式），但这两个不能在同一个属性里出现。
```

>**警告：**
>

> >因为 JSX 的特性更接近 JavaScript 而不是 HTML , 所以 React DOM 使用 camelCase （小驼峰命名）来定义属性的名称，而不是使用 HTML 的属性名称。

Reviewer: ```suggestion
>因为 JSX 的特性更接近 JavaScript 而不是 HTML , 所以 React DOM 使用 `camelCase`（小驼峰命名）来定义属性的名称，而不是使用 HTML 的属性名称。
```

>

> >例如，JSX 里的 class 变成了 className (https://developer.mozilla.org/en-US/docs/Web/API/Element/className)，而 tabindex 则对应着 tabIndex (https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/tabIndex)。

Reviewer: ```suggestion
>例如，JSX 里的 `class` 变成了 [`className`](https://developer.mozilla.org/en-US/docs/Web/API/Element/className)，而 `tabindex` 则对应着 [`tabIndex`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/tabIndex)。
```

### 使用 JSX 指定子项

假如一个标签里面没有内容，你可以把它当作 XML ，在末尾加上 `/>` 来关上它。

```js
const element = <img src={user.avatarUrl} />;
```

JSX 标签里能够包含很多子项:

```js
const element = (

> <div>

Reviewer: ```suggestion
  <div>
```

<h1>Hello!</h1>
<h2>Good to see you here.</h2>
</div>
);
```

### JSX 防止注入攻击

你可以放心地在 JSX 当中使用用户输入：

```js
const title = response.potentiallyMaliciousInput;
// 直接使用是安全的：
const element = <h1>{title}</h1>;
```

> React DOM 在渲染之前默认会过滤 (http://stackoverflow.com/questions/7381974/which-characters-need-to-be-escaped-on-html)  所有传入的值。它可以确保你的应用里没有写进去的信息无法被进行注入攻击。所有的内容在渲染之前都被转换成了字符串。这样可以有效地防止 XSS (跨站脚本) 攻击。

Reviewer: ```suggestion
React DOM 在渲染之前默认会[过滤](http://stackoverflow.com/questions/7381974/which-characters-need-to-be-escaped-on-html)所有传入的值。它可以确保你的应用里没有写进去的信息无法被进行注入攻击。所有的内容在渲染之前都被转换成了字符串。这样可以有效地防止 [XSS (跨站脚本)](https://en.wikipedia.org/wiki/Cross-site_scripting)攻击。
```

### JSX 代表对象（Objects）

Babel 转译器会把 JSX 转换成一个名为 React.createElement() 的方法调用。

下面两种代码的作用是完全相同的：

```js
const element = (

> <h1 className="greeting">

Reviewer: ```suggestion
  <h1 className="greeting">
```

> Hello, world!

Reviewer: ```suggestion
    Hello, world!
```

> </h1>

Reviewer: ```suggestion
  </h1>
```

);
```

```js
const element = React.createElement(

> 'h1',

Reviewer: ```suggestion
  'h1',
```

> {className: 'greeting'},

Reviewer: ```suggestion
  {className: 'greeting'},
```

> 'Hello, world!'

Reviewer: ```suggestion
  'Hello, world!'
```

);
```

> React.createElement() 这个方法首先会进行一些避免 bug 的检查，但更主要的是返回一个类似下面例子的对象：

Reviewer: ```suggestion
`React.createElement()` 这个方法首先会进行一些避免 bug 的检查，但更主要的是返回一个类似下面例子的对象：
```

```js
// Note: this structure is simplified
const element = {
type: 'h1',

> props: {

Reviewer: ```suggestion
  props: {
```

> className: 'greeting',

Reviewer: ```suggestion
    className: 'greeting',
```

> children: 'Hello, world!'

Reviewer: ```suggestion
    children: 'Hello, world!'
```

> }

Reviewer: ```suggestion
  }
```

---
id: introducing-jsx
title: JSX 简介
permalink: docs/introducing-jsx.html
prev: hello-world.html
next: rendering-elements.html
---

> 我们来观察一下声明的这个变量：

Reviewer 2: 我们来观察一下声明的这个变量 => 考虑如下变量声明

用词方面需要注意，不要强行加入自己的意思进去
consider 就是考虑，没有观察的意思
variable declaration 变量声明
这样翻译很自然

Author: 可以

Reviewer: 「考虑如下方式声明变量」如何？

```js
const element = <h1>Hello, world!</h1>;
```

> 这个语法标记看起来很奇怪，因为它既不是字符串也不是 HTML。

Reviewer 2: tag syntax => 标记语法/标签语法
这里不知道为什么要倒装翻译

Author: ok, will change it.

> 它被称为 JSX， 是基于JavaScript 语法的一种扩展。 我们推荐大家用它来描述用户界面。JSX 可能乍一看像模版语言，但它具有JavaScript 的全部功能。

Reviewer: 「基于JavaScript 语法的一种扩展」 ->「基于 JavaScript 语法的一种扩展」

> 它被称为 JSX，是基于 JavaScript 语法的一种扩展。我们推荐大家用它来描述用户界面。JSX 可能乍一看像模版语言，但它具有 JavaScript 的全部功能。

Reviewer 2: It is called JSX, and it is a syntax extension to JavaScript.
它被称为 JSX，是一个 JavaScript 语法扩展。

We recommend using it with React
我们建议在 React 中配合使用 JSX
这句强调点在于

to describe what the UI should look like.
这句没有把真正意思翻译出来，背后的意思是：JSX 这种语法很特殊，能很好的呈现用户界面的本来面目
应该翻译为：JSX 可以很好地描述 UI 应该呈现出的本质形式。
或者其他更好的翻译

remind you of => 使人想起，不应该翻译为“乍一看”这种很奇怪的用词。

Author: 这句 “JSX 可以很好地描述 UI 应该呈现出它应有交互的本质形式” 是不是有点拗口？@dear-lizhihua

> JSX 生成 React "元素"。我们将在[下一章节](/docs/rendering-elements.html) 中探索如何将这些元素渲染到 DOM 里。 下面，我们来看一看 JSX 的基本使用方法，以帮助您入门。

Reviewer 2: 探索 => 探讨/研究，这里用词不当，阅读文档，而不需要探索。

Below, you can find the basics of JSX necessary to get you started.
necessary => 所需，没翻译出来
我们来看下学习 JSX 所需的基础知识

Author: 你们觉得去掉 "以帮助您入门” 怎么样？to get you started 本身就是开始的意思
把最后一句话翻译成 "下面，我们来看一下开始学习 JSX 的所需的基础知识"? @dear-lizhihua

> JSX 生成 React "元素"。我们将在 [下一章节]（/ docs / rendering-elements.html）中探索如何将这些元素渲染到 DOM 里。 下面，我们来看一看 JSX 的基本使用方法，以帮助您入门。

Reviewer: 建议你看看 https://github.com/sparanoid/chinese-copywriting-guidelines ，你的排版有问题。
你重新排版之后我在看吧。

Reviewer: 你看 `[下一章节](/docs /rendering-elements.html)` 这是 md 文件的链接写法，他们是一体的
